File: RxFirebase.java
    @Nullable
    public static Observable<DataSnapshot> observe(final Query query) {
        return Observable.create(new Action1<Emitter<DataSnapshot>>() {
            @Override
            public void call(final Emitter<DataSnapshot> dataSnapshotEmitter) {
                final ValueEventListener valueEventListener = query.addValueEventListener(new ValueEventListener() {
                    @Override
                    public void onDataChange(DataSnapshot dataSnapshot) {
                        dataSnapshotEmitter.onNext(dataSnapshot);
                    }

                    @Override
                    public void onCancelled(DatabaseError databaseError) {
                        dataSnapshotEmitter.onNext(null);
                    }
                });
                dataSnapshotEmitter.setCancellation(new Cancellable() {
                    @Override
                    public void cancel() throws Exception {
                        query.removeEventListener(valueEventListener);
                    }
                });
            }
        }, Emitter.BackpressureMode.BUFFER)
                .observeOn(Schedulers.computation());
    }

File: rxfirebase.kt
fun Query.observe(): Observable<DataSnapshot?> {
    return Observable.create<DataSnapshot?>({
        val listener = this.addValueEventListener(object : ValueEventListener {
            override fun onCancelled(databaseError: DatabaseError) {
                it.onNext(null)
            }

            override fun onDataChange(dataSnapshot: DataSnapshot) {
                it.onNext(dataSnapshot)
            }
        })
        it.setCancellation {
            this.removeEventListener(listener)
        }
    }, Emitter.BackpressureMode.BUFFER)
            .observeOn(Schedulers.computation())
}
